### ISSUE TITLE
Question: Memory control on setting ItemsRepeater's VerticalCacheLength control

### ISSUE THREAD

**Issue by hupo376787** [microsoft/microsoft-ui-xaml#1699]
Hi, I'm designing an app like NetFlix's home page, and using `ItemsRepeater`.

As I has many horizontal repeaters(20+) in an vertical repeater, like the pic shows.
![image](https://user-images.githubusercontent.com/14960168/69856670-2499b100-12c9-11ea-803b-121f309a2468.png)

If I set `VerticalCacheLength` to 30, scroll up and down several times, the app's memory is about 200M all the time.
![image](https://user-images.githubusercontent.com/14960168/69856756-4f840500-12c9-11ea-9d0f-171205cc7a46.png)
If I don't set `VerticalCacheLength` , scroll up and down several times, the app's memory may up to 600M., even more!!!
At the same time, the app react slowly.
After a minutes later, the memory is down to 300M.
![image](https://user-images.githubusercontent.com/14960168/69856856-8f4aec80-12c9-11ea-809a-c93998071972.png)

So I'm wondering the mechanism of `CacheLength`.

**Comment by jevansaks** [comment]
@hupo376787 Can you attach a repro project that shows this? Or maybe rephrase your question? What are you wondering?

**Comment by ranjeshj** [comment]
CacheLength describes the number of viewports (that you see on screen) that will be realized after the visible part of the screen has been realized. 30 viewports seems excessive. I believe the default is 2 viewports. Also when you have nested repeaters in opposite directions, you have this happening in the outer and inner repeaters. I am guessing that a lot got realized which is reflected by the large memory usage and sluggishness is because layout is taking a long time due to the large number of elements in the tree. Can you start with CacheLength of 0 (both horizontal and vertical for both the outer and inner repeaters). Then you can try increasing slowly to find something that works best for your scenario.

**Comment by hupo376787** [comment]
Hi, I'm post this issue, because the `ItemsRepeater` control behaves not very smooth on Xbox One S. I'm design a home page like NetFlix, with vertical 20+ items, each item has 20+ items inside.

Each card's template is a button, with picture inside, and some extra info text.

Actually, we have a Banner(FLipView) at the top of home page, and a media player to play some trailers.

So, these all stuffs generally behaves very slow on Xbox one S. You can just see the slow action while moving focus vertically, and the focus rectangle appears step by step.
While, when scroll horizontally, the focus moves very smoothly.

I have wrote a demo for you guys to test the performance.
Here is the a reporducable [repo](https://github.com/hupo376787/ItemsRepeaterPerformanceTestDemo), and special thanks to @XamlBrewer.

At last, we sincerely looking forward to your reply, and we working together to help make it better. If this is done, our app will be in store soon for xbox and pc users.

![image](https://user-images.githubusercontent.com/14960168/70885981-62127400-2015-11ea-9c41-cbbf39ac5bd0.png)

**Comment by hupo376787** [comment]
[![ImageEx causes performance issue on Xbox One S](https://i.ibb.co/hFQ4Gqp/https-i-ytimg-com-vi-o-FLVUq7p9o8-hqdefault.jpg)](https://youtu.be/oFLVUq7p9o8 "ImageEx causes performance issue on Xbox One S")

**Comment by ranjeshj** [comment]
Hello. I took a quick look at the sample app and think two minor changes (shown in commit below) might help. 
https://github.com/ranjeshj/ItemsRepeaterPerformanceTestDemo/commit/bf73d8eefd4200bf19327f341696d86cfadbff92 

Setting HorizontalCacheLength=1 and VerticalCacheLength=1 on the appropriate ItemsRepeaters brought down the number of elements realized by 30% which should reduce the time taken in layout during keyboard navigation. 

Also, I updated the image source binding to happen in phase 1. Note that for x:Phase to work you currently need to build for .net native. 

I have not tested these changes on an xbox so I cannot tell how much of a difference it will make but it should help a bit.  If you still see perf issues with keyboard navigation, I would suggest getting perf traces and finding out what is taking time.

Thanks,
Ranjesh

**Comment by hupo376787** [comment]
This does not improve effectly. The focus rectangle is still crawling slowly.
My original code has 439 elements, and I used chche length and phase, and scroll up/down several times, the elements is 433.

And a weird thing is, I build & run several times, the result is different. Sometimes maybe 600+.

![image](https://user-images.githubusercontent.com/14960168/70960567-5dec6200-20ba-11ea-8aae-09e85446117a.png)
![image](https://user-images.githubusercontent.com/14960168/70960783-f97dd280-20ba-11ea-8285-f12f44bec71b.png)


Btw, I think this perf is caused by `ImageEx`, I have reported to them, but no reply yet.
If replace to `Image`, perf can be accepted, but no place holder. So I'll not use it currently.

Here is my performance diagsession. One with Cahce Length, the other not.

https://github.com/hupo376787/ItemsRepeaterPerformanceTestDemo/releases/tag/diagsession

**Comment by ranjeshj** [comment]
Are you using ImageEx specifically for the placeholder ? If so, you could put the placeholder in phase 0 and the actual image in phase 1. That way you will get the placeholder behavior when scrolling/loading in the ItemsRepeater with Image.

**Comment by hupo376787** [comment]
Hi, @ranjeshj [ImageEx](https://github.com/windows-toolkit/WindowsCommunityToolkit/tree/master/Microsoft.Toolkit.Uwp.UI.Controls/ImageEx) seems not provide x:phase for place holder.

```
<controls:ImageEx
     PlaceholderStretch="Uniform"
     PlaceholderSource="{Binding is_movie, Converter={StaticResource imageSourceConverter}}"
     Stretch="Uniform"
     Source="{x:Bind cover_image_url, Converter={StaticResource imageUrlCompress}}"/>
```

**Comment by ranjeshj** [comment]
x:Phase is an ItemsRepeater feature (similar to what is described here - https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/optimize-gridview-and-listview#update-items-incrementally). In the DataTemplate provided to ItemsRepeater you can have multiple elements and bindings. x:Phase will help decide what to show first and then phase the rest in the order of the number.

So for example if you had this.
<DataTemplate>
    <Border ... >
       <Image x:Phase='1'  Source=x:bind.. >

When ItemsRepeater loads the item, it will show the Border right away, but bind the image afterwards. The Border here is the placeholder. Any element in the template which does not have x:Phase on it will be shown immediately. This is important because when you are scrolling quickly, we might just show the placeholders and never get to phase 1 until either scrolling slows down or stops. This should provide a better scrolling speed.

**Comment by hupo376787** [comment]
Hi, I think x:phase doesn't improve a lot. The Windows Community Toolkit team has noticed this performace on Xbox One S. https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3080

**Comment by ranjeshj** [comment]
@hupo376787 The number of elements in the template play a huge role in the template especially the elements that are realized and bound. Because they get repeated so many times, reducing a few elements in the template goes a long way. Some ways to reduce them that i can think of 
1. Use RelativePanel instead of multiple nested panels. I see in your sample template that you have 3 textblocks, 2 grids, stackpanel, button and Image. You could use one RelativePanel instead of 2 grids and a stackPanel.
2. Make sure that your text blocks are using fast path - https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/text-block#performance-considerations
3. Do as little work as possible in phase 0. The images, text and anything else should be in x:Phase=1 or later. 
4. ItemsRepeater uses XYFocus Keyboard navigation to put focus on the next element in the direction you are navigating. This is great but does involve figuring out where to go based on element layout positions. If there are a lot of elements - this can become a performance issue. The steps above can help a little with that. Since these are stack layouts, you could also potentially write custom keyboard navigation logic that is based on the data indices would likely be more performant.

**Comment by hupo376787** [comment]
Hi, @ranjeshj . Thanks for the precious advice. I run the fixed demo on Xbox One S, the focus rectangle still moves slowly when scroll up.

Also I noticed, when scrolling down `ItemsRepeater`, there is no scroll animation on my xbox. But when scrolling up, the scroll animation occurs. On my pc, both up and down don't have scroll animation. 
But my production app have scroll animation up & down.

| animation | up | down |
| -- | -- | -- |
| Xbox One S | Yes | No |
| My PC | No | No |

So the scroll animation is also a mystery to me.

Another question is, the fisrt time I scroll down using keyboard/controller, the focus always go to the firsrt element of each horizontal repeater.
Like the picture says, whenever my focus is at 1, 2, 3, 4. they always go to the first.

![image](https://user-images.githubusercontent.com/14960168/71333722-12342f80-2576-11ea-9b36-c635e741104f.png)

On android tv and ios tv, the focus just find the next element at the scroll direction like below.
![image](https://user-images.githubusercontent.com/14960168/71333787-663f1400-2576-11ea-85fc-4c368bda3c4d.png)



Should I open another 2 issues about scroll animation and focus?

**Comment by chingucoding** [comment]
> Like the picture says, whenever my focus is at 1, 2, 3, 4. they always go to the first.

Yes this is because the focus switches from ItemsRepeater to another. Since the ItemsRepeater do not know which index was selected in the other ItemsRepeater, they have no way to select the corresponding item in their own item list.

To have the behavior you described in the second screenshot, other Controls such as GridView may be more suitable for this.

**Comment by hupo376787** [comment]
@chingucoding OK, got it. But how can android & apple can do it. A page like NetFLix home page, what control do they use? The same mechanism?

**Comment by chingucoding** [comment]
I don't know what control(s) they use, especially since they are probably are using Android/IOS controls. Something that they (and you) might do is to achieve this effect is setting the focus manually when the user navigates the page.

**Comment by hupo376787** [comment]
It seems that I have to write code, not using `FindNextFocusableElement` when up/down, because `FindNextFocusableElement` always return the first.

**Comment by chingucoding** [comment]
Yes you would need to look at the (previously) selected index, and do some calculation/logic based on that.

**Comment by ranjeshj** [comment]
@hupo376787 Correct. As i mentioned in the comment above the FindNextFocusableElement call is just to show how to do the manual focus navigation on key down. You can find out which element to put focus on and write the logic for it as you want it. Also, it only works for down because I only coded it for down key :) to see the difference. Down is using the custom logic in KeyDown handler and Up is using XYFocus logic from the framework which has less information to work on and has to look at all the elements on screen to figure out where to go next.

**Comment by hupo376787** [comment]
OK, got it. How about the scroll animation?
Do you have a XBox one S to test the demo?

**Comment by ranjeshj** [comment]
@hupo376787 I am currently on vacation and don't have an XBoxOne S to test it out. In the mean time, you could try calling BringIntoView api on the element you are setting focus on and I believe that has an option to animate. https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.startbringintoview

**Comment by hupo376787** [comment]
I think doing manual navigation stuff helps little on performance. As I said, I'm designing a page more complex than Netflix's home page. I will change my template following your advice, but the main work  belongs to control. We have the same design on android and apple tv, they all run perfect and smooth. Maybe XBox One S is too weak, but Xbox one seems not much stronger than S. So I think WinUI and Community Toolkit can work together, to improve uwp controls.

Btw, Merry Christmas : )

**Comment by ranjeshj** [comment]
hi @hupo376787, were you able to resolve this issue with the manual navigation ? Thanks.

**Comment by ranjeshj** [comment]
Closing this issue since it looks like handling key input has resolved it.
